### PR TITLE
swipe ALL the things!

### PR DESCRIPTION
**Pull request by notwaldorf**
I ported the old `PolymerLabs/swipeable-card` to a magical container that makes all its children swipeable. I tried to comment all the magical animation stuff that was really scary -- @frankiefu, would you mind double checking that I didn't tell any lies in `_animate`?

Features:
- curved and horizontal swiping
- the css transform can be configured, as well as the animation/opacity duration
- all children can be swiped
- swiping can be prevented on the whole container
- swiping can be prevented on individual children

### CODE REVIEW

File: iron-swipeable-container.html [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    _addListeners: function(node) {
+      if (node.nodeType === Node.TEXT_NODE)
+        return;
+      // Set up the animation.
+      node.style.transitionProperty = this.transitionProperty;
+      node.style.transition = node.style.webkitTransition = this.transition;
+
+      this.listen(node, 'track', '_onTrack');
+      this.listen(node, 'transitionend', '_onTransitionEnd');
+      this.listen(node, 'webkitTransitionEnd', '_onTransitionEnd');
```
Comment (frankiefu) [inline iron-swipeable-container.html]: We don't need `webkit` prefix for the `transition`.  All the supporting browsers should work fine without the prefix.  So we can remove this line and all the other `webkitTransition` stuffs.

File: iron-swipeable-container.html [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        // figure out the rotation and vertical displacement needed to
+        // achieve that.
+        var y = totalDistance - Math.sqrt(totalDistance * totalDistance - opaqueDistance * opaqueDistance);
+        var deg = (1 - opacity) * direction * 90;
+
+        translate = 'translate3d(' + x + 'px,' + y + 'px,0)';
+        rotate = ' rotate(' + deg + 'deg)';
+      }
+
+      target.style.transform = target.style.webkitTransform = translate + rotate;
```
Comment (frankiefu) [inline iron-swipeable-container.html]: Suggest to use `this.transform(translate + rotate, target);`

File: iron-swipeable-container.html [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+
+<link rel="import" href="../polymer/polymer.html">
+
+<!--
+`<iron-swipeable-container>` is a container that allows any of its nested
```
Comment (cdata) [inline iron-swipeable-container.html]: :+1:

File: iron-swipeable-container.html [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+<script>
+
+  Polymer({
+    is: 'iron-swipeable-container',
+
+    /**
+     * Fired when a child element is swiped away.
+     *
+     * @event iron-swipe-away
```
Comment (cdata) [inline iron-swipeable-container.html]: Is there another way to swipe something?

File: iron-swipeable-container.html [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+     */
+
+    properties: {
+      /**
+       * The style in which to swipe the card. Currently supported
+       * options are `curve | horizontal`. If left unspecified, the default
+       * is assumed to be `curve`.
```
Comment (cdata) [inline iron-swipeable-container.html]: This is a :bike: :house:: but I have never seed a "curve" swipe IRL. Do you have any examples of where it might be used in practice? Does it make sense as the default?

File: iron-swipeable-container.html [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+       */
+      swipeStyle: {
+       type: String,
+       value: 'curve'
+      },
+
+      /**
+       * If true, then the container will not allow swiping.
+       */
+      disableSwipe: {
```
Comment (cdata) [inline iron-swipeable-container.html]: Since `swipe` is everything that this container exists to do, should this just be `disabled`?

File: iron-swipeable-container.html [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        type: Boolean,
+        value: false
+      },
+
+      /**
+       * The ratio of the width of the element that the translation animation
+       * should happen over. For example, if the `widthRatio` is 3, the
+       * animation will take place on a distance 3 times the width of then
```
Comment (cdata) [inline iron-swipeable-container.html]: `then` -> `the`

File: iron-swipeable-container.html [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+       */
+      transition: {
+        type: String,
+        value: '300ms cubic-bezier(0.4, 0.0, 0.2, 1)'
+      },
+
+      /**
+       * The CSS transition property applied while swiping.
+       */
+      transitionProperty: {
```
Comment (cdata) [inline iron-swipeable-container.html]: Why split this one off? If it's useful to specify the `transitionProperty` separately, why not the timing function and duration? Is there an opportunity to make `transition` a computed value combining the other component values?

File: iron-swipeable-container.html [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+    _removeListeners: function(node) {
+      if (node.nodeType === Node.TEXT_NODE)
+        return;
+      this.unlisten(node, 'track', '_onTrack');
+      this.unlisten(node, 'transitionend', '_onTransitionEnd');
+    },
+
+    detached: function() {
+      if (this._nodeObserver) {
```
Comment (cdata) [inline iron-swipeable-container.html]: It might be good to null this out as well.

File: iron-swipeable-container.html [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        Polymer.dom(this.$.content).unobserveNodes(this._nodeObserver);
+      }
+    },
+
+    _onTrack: function(event) {
+      if (this.disableSwipe)
+        return;
+
+      var target = event.currentTarget;
+      if (target.classList.contains('disable-swipe'))
```
Comment (cdata) [inline iron-swipeable-container.html]: Is this to disable swipe on specific children of the container?

File: iron-swipeable-container.html [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        this._trackMove(track, target);
+      } else if (track.state === 'end') {
+        this._trackEnd(track, target);
+      }
+    },
+
+    _trackStart: function(event, target) {
+      // Save the width of the element, so that we don't trigger a style
+      // recalc every time we need it.
+      this._nodeWidth = target.offsetWidth;
```
Comment (cdata) [inline iron-swipeable-container.html]: :+1:

File: iron-swipeable-container.html [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+        var deg = (1 - opacity) * direction * 90;
+
+        translate = 'translate3d(' + x + 'px,' + y + 'px,0)';
+        rotate = ' rotate(' + deg + 'deg)';
+      }
+
+      this.transform(translate + rotate, target);
+    },
+
+    _swipeEnd: function(away, dir, target) {
```
Comment (cdata) [inline iron-swipeable-container.html]: Boolean parameters always learn a funky taste in my mouth. Is there a better way to express this?

File: iron-swipeable-container.html [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+      // Restore the original transition;
+      target.style.transition = this.transition;
+      this.away = away;
+      this.direction = dir ? 'right' : 'left';
```
Comment (cdata) [inline iron-swipeable-container.html]: Are these deliberately public? If so, it might be good to pick names that give more context to their meaning.

Comment (cdata) [inline iron-swipeable-container.html]: (also, it would be good to initialize the values in `created` either way)

File: iron-swipeable-container.html [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+        // figure out the rotation and vertical displacement needed to
+        // achieve that.
+        var y = totalDistance - Math.sqrt(totalDistance * totalDistance - opaqueDistance * opaqueDistance);
+        var deg = (1 - opacity) * direction * 90;
+
+        translate = 'translate3d(' + x + 'px,' + y + 'px,0)';
+        rotate = ' rotate(' + deg + 'deg)';
+      }
+
+      this.transform(translate + rotate, target);
```
Comment (cdata) [inline iron-swipeable-container.html]: :+1:

File: iron-swipeable-container.html [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+      }
+    },
+
+    _onTransitionEnd: function(event) {
+      var target = event.currentTarget;
+
+      if (this.away && event.propertyName === 'opacity') {
+        this.fire('iron-swipe-away', {direction: this.direction, target:target});
+      }
+      event.stopPropagation();
```
Comment (cdata) [inline iron-swipeable-container.html]: Why is propagation stopped?

File: iron-swipeable-container.html [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+    },
+
+    _swipeAway: function(event) {
+      Polymer.dom(this).removeChild(event.detail.target);
```
Comment (cdata) [inline iron-swipeable-container.html]: It seems like it would be better to spare adding a listener for this event, and instead just move this right after 257.

File: iron-swipeable-container.html [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+      <div class="disable-swipe">I cannot be swiped be swiped</div>
+      <paper-card heading="But I can!"></paper-card>
+    </iron-swipeable-container>
+
+    <iron-swipeable-container disable-swipe>
+      <div>I cannot be swiped</div>
+      <paper-card heading="Me neither :("></paper-card>
+    </iron-swipeable-container>
+
+It is a good idea to disable text selection on any of the children that you
```
Comment (cdata) [inline iron-swipeable-container.html]: This is a problem with a lot of nuance, and it might be good to mention that there are limitations to the approach you are proposing (e.g., text in the children will not be selectable). Ideally I think you want to selectively disable `user-select` automatically on behalf of the user when you think there is a swipe taking place, but this is easier described than done..

Comment (cdata) [inline iron-swipeable-container.html]: In Polytrack, I attempted to do this by measuring how much time there was between `down` and the first `track` event to decide if the user intended to drag. If the time is < 50ms, I assume the user intended to drag, otherwise I assume that there was some other desired intention.

File: test/basic.html [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+        var element = container.querySelector('#native');
+
+        var swipeEventHandler = sinon.spy();
+        container.addEventListener('iron-swipe-away', swipeEventHandler);
+
+        Polymer.Base.async(function() {
+          expect(Polymer.dom(container).queryDistributedElements('*').length).to.be.equal(2);
+          MockInteractions.track(element, 60, 0);
+
+          setTimeout(function(){
```
Comment (cdata) [inline test/basic.html]: Is there a specific reason why you used `setTimeout` instead of `Polymer.Base.async` here?

File: test/basic.html [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+    <template>
+      <iron-swipeable-container>
+        <div id="native" class="disable-swipe">Hello</div>
+        <test-element id="custom" class="disable-swipe"></test-element>
+      </iron-swipeable-container>
+    </template>
+  </test-fixture>
+
+
+  <test-fixture id="no-swipe" disable-swipe>
```
Comment (cdata) [inline test/basic.html]: Why does the `test-fixture` have a `disable-swipe` attribute here?

File: test/basic.html [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+-->
+<html>
+<head>
+  <meta charset="UTF-8">
+  <title>iron-swipeable-container tests</title>
+  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0">
+
+  <script src="../../webcomponentsjs/webcomponents-lite.js"></script>
+  <script src="../../web-component-tester/browser.js"></script>
+  <script src="../../test-fixture/test-fixture-mocha.js"></script>
```
Comment (cdata) [inline test/basic.html]: If you require `web-component-tester >= v3.4.0`, you can remove this line.

File: test/basic.html [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+  <script src="../../iron-test-helpers/mock-interactions.js"></script>
+
+  <link rel="import" href="../../test-fixture/test-fixture.html">
```
Comment (cdata) [inline test/basic.html]: If you require `web-component-tester >= v3.4.0`, you can remove this line.

File: iron-swipeable-container.html (hunk 5 above)
Comment (notwaldorf) [inline iron-swipeable-container.html]: That's what @frankiefu had originally as a default swipe, but I guess you're right. Maybe horizontal is a better default.

File: iron-swipeable-container.html (hunk 10 above)
Comment (notwaldorf) [inline iron-swipeable-container.html]: Yup! See the docs :)

File: iron-swipeable-container.html (hunk 5 above)
Comment (notwaldorf) [inline iron-swipeable-container.html]: I double checked with the MD motion interaction designer and he suggested the horizontal to be the default, especially since iOS is the only MD thing that uses the curve animation (we'll keep it, but just flip the boolean)

File: iron-swipeable-container.html (hunk 4 above)
Comment (notwaldorf) [inline iron-swipeable-container.html]: I think i was just being pedantic. Fixed!

File: iron-swipeable-container.html (hunk 8 above)
Comment (notwaldorf) [inline iron-swipeable-container.html]: I think I'm just going to make it private. I think transform and opacity is the only thing we should support.

File: iron-swipeable-container.html (hunk 17 above)
Comment (notwaldorf) [inline iron-swipeable-container.html]: Can I do this in a different PR, maybe? As an enhancement?

File: iron-swipeable-container.html (hunk 9 above)
Comment (notwaldorf) [inline iron-swipeable-container.html]: done!

File: iron-swipeable-container.html (hunk 5 above)
Comment (kbwatts) [inline iron-swipeable-container.html]: Nice one! That makes sense :)

**Comment by valdrinkoshi** [comment]
@notwaldorf any chance for this to be a behavior? It will be handy for other elements like [`paper-toast`](https://github.com/PolymerElements/paper-toast/issues/6)

**Comment by notwaldorf** [comment]
@valdrinkoshi this would have to be in addition to, not instead of, the current implementation. If it's a behaviour, then only things that have implemented the behaviour can be swiped, which is a bit lame. Alternatively, you can have all of the toasts that can appear on a given page live inside a swipeable container, which would make them swipeable. That's actually more what you're looking for, if I think about it more

**Comment by notwaldorf** [comment]
@cdata paging for review!

File: iron-swipeable-container.html [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+      } else if (track.state === 'end') {
+        this._trackEnd(track, target);
+      }
+    },
+
+    _trackStart: function(event, target) {
+      // Save the width of the element, so that we don't trigger a style
+      // recalc every time we need it.
+      this._nodeWidth = target.offsetWidth;
+      target.style.transition = 'none';
```
Comment (valdrinkoshi) [inline iron-swipeable-container.html]: when the swipe starts, we should disable text selection on the target. Consider adding a class to the `iron-swipeable-container`: `this.toggleClass('swiping', true)`
```css
.swiping {
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  user-select: none;
  cursor: default;
}
```
also, remove the class on swipe end too

File: iron-swipeable-container.html [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+      }
+
+      this.transform(translate + rotate, target);
+    },
+
+    _swipeEnd: function(target) {
+      // Restore the original transition;
+      target.style.transition = this.transition;
+
+      if (this._swipeComplete) {
```
Comment (valdrinkoshi) [inline iron-swipeable-container.html]: it would be nice to have an event fired at the end of the transition, to know if the target was removed or not

File: iron-swipeable-container.html [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+        this._animate(0, target);
+      }
+    },
+
+    _onTransitionEnd: function(event) {
+      var target = event.currentTarget;
+
+      if (this._swipeComplete && event.propertyName === 'opacity') {
+        Polymer.dom(this).removeChild(target);
+        this.fire('iron-swipe',
```
Comment (valdrinkoshi) [inline iron-swipeable-container.html]: \o/

File: README.md [hunk 25]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (cdata) [inline README.md]: Isn't this file auto-generated?

Comment (notwaldorf) [inline README.md]: yeah, but my editor still hates on lines with trailing spaces, and i don't think I should stop it :)

File: demo/index.html [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+      padding: 20px;
+      background-color: var(--paper-lime-500);
+    }
+
+    .item.blue {
+      background-color: var(--paper-blue-300);
+    }
+
+    /* If you want your things to swipe nicely, you might want to disable
+     * text selection on those elements. */
```
Comment (cdata) [inline demo/index.html]: protip :100:

File: iron-swipeable-container.html (hunk 17 above)
Comment (cdata) [inline iron-swipeable-container.html]: SGTM

Comment (cdata) [inline iron-swipeable-container.html]: (btw, this is requesting clarification in documentation, not a change in implementation).

File: iron-swipeable-container.html [hunk 27]
```diff
@@ -0,0 +1,261 @@
...
+        return;
+
+      var track = event.detail;
+      if (track.state === 'start') {
+        this._trackStart(track, target);
+      } else if (track.state === 'track') {
+        this._trackMove(track, target);
+      } else if (track.state === 'end') {
+        this._trackEnd(track, target);
+      }
```
Comment (cdata) [inline iron-swipeable-container.html]: I really wish there were just separate events I could listen to for this kind of thing :( /cc @azakus

**Comment by cdata** [comment]
This LGTM :+1: nice work!
